hacks/grav.man
.TH XScreenSaver 1 "10-May-97" "X Version 11"
.SH NAME
grav \- draws a simple orbital simulation
.SH SYNOPSIS
.B grav
[\-\-display \fIhost:display.screen\fP] [\-\-foreground \fIcolor\fP]
[\-\-background \fIcolor\fP] [\-\-window] [\-\-root]
[\-\-window\-id \fInumber\fP][\-\-mono] [\-\-install] [\-\-visual \fIvisual\fP] [\-\-ncolors \fIinteger\fP] [\-\-delay \fImicroseconds\fP] [\-\-count \fIinteger\fP] [\-\-decay] [\-\-no\-decay] [\-\-trail] [\-\-no\-trail]

[\-\-fps]
.SH DESCRIPTION
The \fIgrav\fP program draws a simple orbital simulation
.SH OPTIONS
.I grav
accepts the following options:
.TP 8
.B \-\-window
Draw on a newly-created window.  This is the default.
.TP 8
.B \-\-root
Draw on the root window.
.TP 8
.B \-\-window\-id \fInumber\fP
Draw on the specified window.
.TP 8
.B \-\-mono 
If on a color display, pretend we're on a monochrome display.
.TP 8
.B \-\-install
Install a private colormap for the window.
.TP 8
.B \-\-visual \fIvisual\fP
Specify which visual to use.  Legal values are the name of a visual class,
or the id number (decimal or hex) of a specific visual.
.TP 8
.B \-\-ncolors \fIinteger\fP
How many colors should be used (if possible).  Default 64.
The colors are chosen randomly.
.TP 8
.B \-\-count \fIinteger\fP
Default 12.
.TP 8
.B \-\-decay
.TP 8
.B \-\-no\-decay
Whether orbits should decay.

.TP 8
.B \-\-trail
.TP 8
.B \-\-no\-trail
Whether the objects should leave trails behind them (makes it look vaguely
like a cloud-chamber.

.TP 8
.B \-\-fps
Display the current frame rate and CPU load.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.TP 8
.B XSCREENSAVER_WINDOW
The window ID to use with \fI\-\-root\fP.
.SH SEE ALSO
.BR X (1),
.BR xscreensaver (1),
.BR xlock (1)
.SH COPYRIGHT
Copyright \(co 1993 by Greg Bowering.

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation. 
.SH AUTHOR
Greg Bowering <gb@pobox.com>, 1993.

Ability to run standalone or with \fIxscreensaver\fP added by 
Jamie Zawinski <jwz@jwz.org>, 10-May-97.


po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from
	Guillermo Movia <gmovia@hotpop.com> and converted to UTF-8

2002-04-29  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* wa.po: Updated Walloon file

2002-04-27  Duarte Loreto <happyguy_pt@hotmail.com>

        * pt.po: Added Portuguese translation.

2002-04-26  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

        * et.po: Updated Estonian translation.

2002-04-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-04-21  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmÃ¥l) translation.


hacks/grav.c
/* -*- Mode: C; tab-width: 4 -*- */
/* grav --- planets spinning around a pulsar */

#if 0
static const char sccsid[] = "@(#)grav.c	5.00 2000/11/01 xlockmore";
#endif

/*-
 * Copyright (c) 1993 by Greg Boewring <gb@pobox.com>
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted,
 * provided that the above copyright notice appear in all copies and that
 * both that copyright notice and this permission notice appear in
 * supporting documentation.
 *
 * This file is provided AS IS with no warranties of any kind.  The author
 * shall have no liability with respect to the infringement of copyrights,
 * trade secrets or any patents by this file or any part thereof.  In no
 * event will the author be liable for any lost revenue or profits or
 * other special, indirect and consequential damages.
 *
 * Revision History:
 * 01-Nov-2000: Allocation checks
 * 10-May-1997: Compatible with xscreensaver
 * 11-Jul-1994: color version
 * 06-Oct-1993: Written by Greg Bowering <gb@pobox.com>
 */

#ifdef STANDALONE
#define MODE_grav
#define DEFAULTS	"*delay: 10000 \n" \
					"*count: 12 \n" \
					"*ncolors: 64 \n" \
					"*fpsSolid: true \n" \
					"*ignoreRotation: True \n" \

/*				    "*lowrez: True \n" \ */

#define BRIGHT_COLORS
# define release_grav 0
# define grav_handle_event 0
# include "xlockmore.h"		/* in xscreensaver distribution */
#else /* STANDALONE */
# include "xlock.h"		/* in xlockmore distribution */
#endif /* STANDALONE */

#ifdef MODE_grav

#define DEF_DECAY "True"	/* Damping for decaying orbits */
#define DEF_TRAIL "True"	/* For trails (works good in mono only) */

static Bool decay;
static Bool trail;

static XrmOptionDescRec opts[] =
{
	{"-decay", ".grav.decay", XrmoptionNoArg, "on"},
	{"+decay", ".grav.decay", XrmoptionNoArg, "off"},
	{"-trail", ".grav.trail", XrmoptionNoArg, "on"},
	{"+trail", ".grav.trail", XrmoptionNoArg, "off"}
};
static argtype vars[] =
{
	{&decay, "decay", "Decay", DEF_DECAY, t_Bool},
	{&trail, "trail", "Trail", DEF_TRAIL, t_Bool}
};
static OptionStruct desc[] =
{
	{"-/+decay", "turn on/off decaying orbits"},
	{"-/+trail", "turn on/off trail dots"}
};

ENTRYPOINT ModeSpecOpt grav_opts =
{sizeof opts / sizeof opts[0], opts, sizeof vars / sizeof vars[0], vars, desc};

#ifdef USE_MODULES
ModStruct   grav_description =
{"grav", "init_grav", "draw_grav", (char *) NULL,
 "refresh_grav", "init_grav", "free_grav", &grav_opts,
 10000, -12, 1, 1, 64, 1.0, "",
 "Shows orbiting planets", 0, NULL};

#endif

#define GRAV			-0.02	/* Gravitational constant */
#define DIST			16.0
#define COLLIDE			0.0001
#define ALMOST			15.99
#define HALF			0.5
/* #define INTRINSIC_RADIUS     200.0 */
#define INTRINSIC_RADIUS	((float) (gp->height/5))
#define STARRADIUS		(unsigned int)(gp->height/(2*DIST))
#define AVG_RADIUS		(INTRINSIC_RADIUS/DIST)
#define RADIUS			(unsigned int)(INTRINSIC_RADIUS/(POS(Z)+DIST))

#define XR			HALF*ALMOST
#define YR			HALF*ALMOST
#define ZR			HALF*ALMOST

#define VR			0.04

#define DIMENSIONS		3
#define X			0
#define Y			1
#define Z			2

#define DAMP			0.999999
#define MaxA			0.1	/* Maximum acceleration (w/ damping) */

#define POS(c) planet->P[c]
#define VEL(c) planet->V[c]
#define ACC(c) planet->A[c]

#define Planet(x,y)\
  if ((x) >= 0 && (y) >= 0 && (x) <= gp->width && (y) <= gp->height) {\
     XFillArc(display, window, gc,\
      (x) - planet->ri / 2, (y) - planet->ri / 2, planet->ri, planet->ri,\
      0, 23040);\
   }

#define FLOATRAND(min,max)	((min)+(LRAND()/MAXRAND)*((max)-(min)))

typedef struct {
	double      P[DIMENSIONS], V[DIMENSIONS], A[DIMENSIONS];
	int         xi, yi, ri;
	unsigned long colors;
} planetstruct;

typedef struct {
	int         width, height;
	int         x, y, sr, nplanets;
	unsigned long starcolor;
	planetstruct *planets;
} gravstruct;

static gravstruct *gravs = (gravstruct *) NULL;

static void
init_planet(ModeInfo * mi, planetstruct * planet)
{
	gravstruct *gp = &gravs[MI_SCREEN(mi)];

# ifdef HAVE_JWXYZ
    jwxyz_XSetAntiAliasing (MI_DISPLAY(mi), MI_GC(mi), False);
# endif

	if (MI_NPIXELS(mi) > 2)
		planet->colors = MI_PIXEL(mi, NRAND(MI_NPIXELS(mi)));
	else
		planet->colors = MI_WHITE_PIXEL(mi);
	/* Initialize positions */
	POS(X) = FLOATRAND(-XR, XR);
	POS(Y) = FLOATRAND(-YR, YR);
	POS(Z) = FLOATRAND(-ZR, ZR);

	if (POS(Z) > -ALMOST) {
		planet->xi = (int)
			((double) gp->width * (HALF + POS(X) / (POS(Z) + DIST)));
		planet->yi = (int)
			((double) gp->height * (HALF + POS(Y) / (POS(Z) + DIST)));
	} else
		planet->xi = planet->yi = -1;
	planet->ri = RADIUS;

	/* Initialize velocities */
	VEL(X) = FLOATRAND(-VR, VR);
	VEL(Y) = FLOATRAND(-VR, VR);
	VEL(Z) = FLOATRAND(-VR, VR);
}

static void
draw_planet(ModeInfo * mi, planetstruct * planet)
{
	Display    *display = MI_DISPLAY(mi);
	Window      window = MI_WINDOW(mi);
	GC          gc = MI_GC(mi);
	gravstruct *gp = &gravs[MI_SCREEN(mi)];
	double      D;		/* A distance variable to work with */
	register unsigned char cmpt;

	D = POS(X) * POS(X) + POS(Y) * POS(Y) + POS(Z) * POS(Z);
	if (D < COLLIDE)
		D = COLLIDE;
	D = sqrt(D);
	D = D * D * D;
	for (cmpt = X; cmpt < DIMENSIONS; cmpt++) {
		ACC(cmpt) = POS(cmpt) * GRAV / D;
		if (decay) {
			if (ACC(cmpt) > MaxA)
				ACC(cmpt) = MaxA;
			else if (ACC(cmpt) < -MaxA)
				ACC(cmpt) = -MaxA;
			VEL(cmpt) = VEL(cmpt) + ACC(cmpt);
			VEL(cmpt) *= DAMP;
		} else {
			/* update velocity */
			VEL(cmpt) = VEL(cmpt) + ACC(cmpt);
		}
		/* update position */
		POS(cmpt) = POS(cmpt) + VEL(cmpt);
	}

	gp->x = planet->xi;
	gp->y = planet->yi;

	if (POS(Z) > -ALMOST) {
		planet->xi = (int)
			((double) gp->width * (HALF + POS(X) / (POS(Z) + DIST)));
		planet->yi = (int)
			((double) gp->height * (HALF + POS(Y) / (POS(Z) + DIST)));
	} else
		planet->xi = planet->yi = -1;

	/* Mask */
	XSetForeground(display, gc, MI_BLACK_PIXEL(mi));
	Planet(gp->x, gp->y);
	if (trail) {
        int r = 1;
        if (gp->width > 2560 || gp->height > 2560)
          r *= 3;  /* Retina displays */
		XSetForeground(display, gc, planet->colors);
        XFillArc(display, MI_WINDOW(mi), gc,
                 gp->x-r/2, gp->y-r/2, r, r,
                 0, 23040);
	}
	/* Move */
	gp->x = planet->xi;
	gp->y = planet->yi;
	planet->ri = RADIUS;

	/* Redraw */
	XSetForeground(display, gc, planet->colors);
	Planet(gp->x, gp->y);
}

ENTRYPOINT void
init_grav(ModeInfo * mi)
{
	unsigned char ball;
	gravstruct *gp;

	MI_INIT (mi, gravs);
	gp = &gravs[MI_SCREEN(mi)];

	gp->width = MI_WIDTH(mi);
	gp->height = MI_HEIGHT(mi);

	gp->sr = STARRADIUS;

	gp->nplanets = MI_COUNT(mi);
	if (gp->nplanets < 0) {
		if (gp->planets) {
			(void) free((void *) gp->planets);
			gp->planets = (planetstruct *) NULL;
		}
		gp->nplanets = NRAND(-gp->nplanets) + 1;	/* Add 1 so its not too boring */
	}
	if (gp->planets == NULL) {
		if ((gp->planets = (planetstruct *) calloc(gp->nplanets,
				sizeof (planetstruct))) == NULL)
			return;
	}

	MI_CLEARWINDOW(mi);

	if (MI_NPIXELS(mi) > 2)
		gp->starcolor = MI_PIXEL(mi, NRAND(MI_NPIXELS(mi)));
	else
		gp->starcolor = MI_WHITE_PIXEL(mi);
	for (ball = 0; ball < (unsigned char) gp->nplanets; ball++)
		init_planet(mi, &gp->planets[ball]);
}

ENTRYPOINT void
draw_grav(ModeInfo * mi)
{
	Display    *display = MI_DISPLAY(mi);
	Window      window = MI_WINDOW(mi);
	GC          gc = MI_GC(mi);
	register unsigned char ball;
	gravstruct *gp;

	if (gravs == NULL)
			return;
	gp = &gravs[MI_SCREEN(mi)];
	if (gp->planets == NULL)
		return;

	if (!MI_IS_DRAWN(mi)) {
		for (ball = 0; ball < (unsigned char) gp->nplanets; ball++) {
			planetstruct *planet = &gp->planets[ball];

			/* Draw planets */
			Planet(planet->xi, planet->yi);
		}

		/* Draw centrepoint */
		XDrawArc(display, MI_WINDOW(mi), gc,
			 gp->width / 2 - gp->sr / 2, gp->height / 2 - gp->sr / 2, gp->sr, gp->sr,
			 0, 23040);
	}

	MI_IS_DRAWN(mi) = True;
	/* Mask centrepoint */
	XSetForeground(display, gc, MI_BLACK_PIXEL(mi));
	XDrawArc(display, window, gc,
		 gp->width / 2 - gp->sr / 2, gp->height / 2 - gp->sr / 2, gp->sr, gp->sr,
		 0, 23040);

	/* Resize centrepoint */
	switch (NRAND(4)) {
		case 0:
			if (gp->sr < (int) STARRADIUS)
				gp->sr++;
			break;
		case 1:
			if (gp->sr > 2)
				gp->sr--;
	}

	/* Draw centrepoint */
	XSetForeground(display, gc, gp->starcolor);
	XDrawArc(display, window, gc,
		 gp->width / 2 - gp->sr / 2, gp->height / 2 - gp->sr / 2, gp->sr, gp->sr,
		 0, 23040);

	for (ball = 0; ball < (unsigned char) gp->nplanets; ball++)
		draw_planet(mi, &gp->planets[ball]);
}

ENTRYPOINT void
reshape_grav(ModeInfo * mi, int width, int height)
{
	gravstruct *gp = &gravs[MI_SCREEN(mi)];
	gp->width  = width;
	gp->height = height;
    XClearWindow (MI_DISPLAY (mi), MI_WINDOW(mi));
}

ENTRYPOINT void
free_grav(ModeInfo * mi)
{
	gravstruct *gp = &gravs[MI_SCREEN(mi)];
	if (gp->planets)
		(void) free((void *) gp->planets);
}

#ifndef STANDALONE
ENTRYPOINT void
refresh_grav(ModeInfo * mi)
{
	MI_CLEARWINDOW(mi);
}
#endif

XSCREENSAVER_MODULE ("Grav", grav)

#endif /* MODE_grav */
